src/command/restore/restore.c

```diff
@@ -0,0 +1,1744 @@
+
+                        strCat(result, PAUSE_AT_RECOVERY_TARGET " = 'false'\n");
+                    }
+                    // Else error on unsupported version
+                    else
+                    {
+                        THROW_FMT(
+                            OptionInvalidError, CFGOPT_TARGET_ACTION " option is only available in PostgreSQL >= %s",
+                            strPtr(pgVersionToStr(PG_VERSION_RECOVERY_TARGET_PAUSE)));
+                    }
+                }
+            }
+
+            // Write recovery_target_timeline
+            if (cfgOptionTest(cfgOptTargetTimeline))
+                strCatFmt(result, RECOVERY_TARGET_TIMELINE " = '%s'\n", strPtr(cfgOptionStr(cfgOptTargetTimeline)));
+
+            memContextSwitch(MEM_CONTEXT_OLD());
+            result = strDup(result);
+            memContextSwitch(MEM_CONTEXT_TEMP());
+        }
+        MEM_CONTEXT_TEMP_END();
+    }
+
+    FUNCTION_LOG_RETURN(STRING, result);
+}
+
+/***********************************************************************************************************************************
+Log the results of a job and throw errors
+***********************************************************************************************************************************/
+// Helper function to determine if a file should be zeroed
+static bool
+restoreFileZeroed(const String *manifestName, RegExp *zeroExp)
+{
+    FUNCTION_TEST_BEGIN();
+        FUNCTION_TEST_PARAM(STRING, manifestName);
+        FUNCTION_TEST_PARAM(REGEXP, zeroExp);
+    FUNCTION_TEST_END();
+
+    ASSERT(manifestName != NULL);
+
+    FUNCTION_TEST_RETURN(
+        zeroExp == NULL ? false : regExpMatch(zeroExp, manifestName) && !strEndsWith(manifestName, STRDEF("/" PG_FILE_PGVERSION)));
+}
+
+// Helper function to construct the absolute pg path for any file
+static String *
+restoreFilePgPath(const Manifest *manifest, const String *manifestName)
+{
+    FUNCTION_TEST_BEGIN();
+        FUNCTION_TEST_PARAM(MANIFEST, manifest);
+        FUNCTION_TEST_PARAM(STRING, manifestName);
+    FUNCTION_TEST_END();
+
+    ASSERT(manifest != NULL);
+    ASSERT(manifestName != NULL);
+
+    String *result = strNewFmt("%s/%s", strPtr(manifestTargetBase(manifest)->path), strPtr(manifestPgPath(manifestName)));
+
+    if (strEq(manifestName, STRDEF(MANIFEST_TARGET_PGDATA "/" PG_PATH_GLOBAL "/" PG_FILE_PGCONTROL)))
+        result = strNewFmt("%s." STORAGE_FILE_TEMP_EXT, strPtr(result));
+
+    FUNCTION_TEST_RETURN(result);
+}
+
+static uint64_t
+restoreJobResult(const Manifest *manifest, ProtocolParallelJob *job, RegExp *zeroExp, uint64_t sizeTotal, uint64_t sizeRestored)
+{
+    FUNCTION_LOG_BEGIN(logLevelDebug);
+        FUNCTION_LOG_PARAM(MANIFEST, manifest);
+        FUNCTION_LOG_PARAM(PROTOCOL_PARALLEL_JOB, job);
+        FUNCTION_LOG_PARAM(REGEXP, zeroExp);
+        FUNCTION_LOG_PARAM(UINT64, sizeTotal);
+        FUNCTION_LOG_PARAM(UINT64, sizeRestored);
+    FUNCTION_LOG_END();
+
+    ASSERT(
```

Comment:
This seems too late. In the original code, we added pg_control into the jobQueue to be copied into destination pg_control.tmp so that we'd never have a valid pg_control until we were successfully done. But now we're not filtering (restoreProcessQueue function uses strLstAdd(targetList, STRDEF(MANIFEST_TARGET_PGDATA "/")); so won't pg_data/global/pg_control pass that test?) which means we'll be copying pg_control during the copy process as the same name and then when all is said and done, we're renaming it. That is too late since now we have a valid pg_control during the process when we shouldn't.

Reply:
Shoot, this comment does not look like it is in the right spot. It should be at the point where 
        // Rename pg_control
        if (storageExistsNP(storagePg(), STRDEF(PG_PATH_GLOBAL "/" PG_FILE_PGCONTROL "." STORAGE_FILE_TEMP_EXT)))

Reply:
pg_control is renamed before copy by `restoreFilePgPath()` in `restoreJobCallback()`.  We also use `restoreFilePgPath()` in `restoreJobResult()` for reporting.

We can see this works because we have an entry in the log:
```
P01   INFO: restore file {[path]}/pg/global/pg_control.pgbackrest.tmp (8KB, 87%) checksum 5e2b96c19c4f5c63a5afa2de504d29fe64a4c908
```
And:
```
P00   INFO: restore global/pg_control (performed last to ensure aborted restores cannot be started)
```
if `global/pg_control.pgbackrest.tmp` did not exist we would see:
```
P00   WARN: backup does not contain 'global/pg_control' -- cluster will not start
```